### PR TITLE
feat: dp rank routing

### PR DESCRIPTION
**Pull request by PeaBrane**
#### Overview:

- New protocol `WorkerWithDpRank` which would include `worker_id` and `dp_rank`. All core components in `kv_router` would work with this protocol.
- Register `dp_size` as another `RuntimeConfig`

#### Minor Notes

- We need to keep `find_best_match` at the python binding level still return `worker_id` (without the `dp_rank` info), otherwise it would be too breaking for many downstream components (e.g. sglang). Will think about dealing with this in future PRs to keep the scope in check
- We keep `dp_rank` as optional in `PreprocessedRequest`. However, when this reaches the core components in `kv_router` and `mocker`, we would always unwrap it to 0, otherwise the indexer + slot manager logics become too messy. Naturally, this means the publisher would need to publish a non-optional dp_rank (which I believe is already the current behavior)

#### e2e tests

- No regression on original vllm benchmarkings (no dp ranks). 
- e2e test with a vllm/mocker worker with multiple workers and multiple dp ranks and produced sensible caching results.

## Summary by CodeRabbit

- New Features
  - Added data-parallel execution support. You can include an optional data-parallel rank in requests, and set data_parallel_size in runtime configuration. Routing and scheduling now leverage this for multi-GPU setups.
- Improvements
  - More accurate worker selection and load aggregation across data-parallel ranks, improving performance under load.
  - KV cache events and related metrics now include the data-parallel rank for clearer observability and troubleshooting.

### CODE REVIEW

File: components/src/dynamo/vllm/handlers.py [hunk 1]
```diff
@@ -201,20 +208,23 @@ async def generate(self, request, context):
...
 
+        # Extract dp_rank from request if available
+        dp_rank = request.get("dp_rank")
+
         async with self._abort_monitor(context, request_id):
             try:
                 async for tok in self.generate_tokens(
-                    prompt, sampling_params, request_id
+                    prompt, sampling_params, request_id, data_parallel_rank=dp_rank
                 ):
```
[the remark this answers is not in the document]
Comment (PeaBrane) [reply components/src/dynamo/vllm/handlers.py]: fine for now

File: lib/llm/src/kv_router/indexer.rs [hunk 2]
```diff
@@ -430,48 +435,72 @@ impl RadixTree {
                 Ok(())
             }
             KvCacheEventData::Cleared => {
-                self.clear_all_blocks(worker_id);
+                self.clear_all_blocks(worker.worker_id);
                 Ok(())
             }
         }
```
[the remark this answers is not in the document]
Comment (PeaBrane) [reply lib/llm/src/kv_router/indexer.rs:438]: this is desired

File: lib/llm/src/kv_router/publisher.rs [hunk 3]
```diff
@@ -326,13 +326,16 @@ pub async fn start_zmq_listener(
...
+                    "ZMQ listener on {} received batch with {} events (seq={}, dp_rank={})",
                     zmq_endpoint,
                     batch.events.len(),
-                    seq
+                    seq,
+                    batch.data_parallel_rank
                 );
+
+                let dp_rank = Some(batch.data_parallel_rank);
                 for raw_event in batch.events.into_iter() {
```
[the remark this answers is not in the document]
Comment (PeaBrane) [reply lib/llm/src/kv_router/publisher.rs:337]: but even before, we are already keeping dp_rank in the batched struct, and this commit did not change that

File: lib/llm/src/kv_router/sequence.rs [hunk 4]
```diff
@@ -591,10 +591,10 @@ impl ActiveSequencesMultiWorker {
...
         isl: usize,
         overlap: u32,
-        worker_id: WorkerId,
+        worker: WorkerWithDpRank,
     ) -> Result<()> {
-        if !self.senders.contains_key(&worker_id) {
-            return Err(anyhow::anyhow!("Worker ID {worker_id} not found"));
+        if !self.senders.contains_key(&worker.worker_id) {
+            return Err(anyhow::anyhow!("Worker ID {} not found", worker.worker_id));
         }
```
[the remark this answers is not in the document]
Comment (PeaBrane) [reply lib/llm/src/kv_router/sequence.rs:619]: fine for now

File: lib/llm/src/protocols/common/preprocessor.rs [hunk 5]
```diff
@@ -65,6 +65,11 @@ pub struct PreprocessedRequest {
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub disaggregated_params: Option<serde_json::Value>,
 
+    /// Data parallel rank for the request (used with data parallelism)
+    #[builder(default)]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub dp_rank: Option<u32>,
+
```
[the remark this answers is not in the document]
Comment (PeaBrane) [reply lib/llm/src/protocols/common/preprocessor.rs:68]: the Builder is derived no?

File: lib/llm/src/kv_router/indexer.rs [hunk 6]
```diff
@@ -332,11 +332,15 @@ impl RadixTree {
...
+        let (worker_id, kv_event) = (event.worker_id, event.event);
+        let (id, op) = (kv_event.event_id, kv_event.data);
+
+        // Construct WorkerWithDpRank from worker_id and dp_rank from the event
+        let worker = WorkerWithDpRank::new(worker_id, kv_event.dp_rank);
+
         tracing::trace!(id, "RadixTree::apply_event: Store operation: {:?}", op);
 
-        let worker_lookup = self.lookup.entry(worker_id).or_default();
+        let worker_lookup = self.lookup.entry(worker).or_default();
```
[the remark this answers is not in the document]
Comment (PeaBrane) [reply lib/llm/src/kv_router/indexer.rs:341]: I think the compiler is smart enough at that arm to drop worker_lookup

File: lib/llm/src/kv_router/publisher.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (PeaBrane) [reply lib/llm/src/kv_router/publisher.rs:337]: if this were a problem we would have known about it by now

**Comment by ishandhanani** [comment]
bro brought it back from the dead

File: lib/llm/src/kv_router/indexer.rs [hunk 7]
```diff
@@ -57,6 +57,9 @@ pub const XXH3_SEED: u64 = 1337;
 use crate::kv_router::protocols::*;
 use crate::tokens::SequenceHash;
 
+// Re-export WorkerId for backward compatibility
+pub type WorkerId = i64;
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/indexer.rs]: Can we move to protocols

File: lib/llm/src/kv_router/indexer.rs [hunk 8]
```diff
@@ -430,48 +435,72 @@ impl RadixTree {
...
         }
     }
 
-    pub fn clear_all_blocks(&mut self, worker: WorkerId) {
-        // Check if the worker has any blocks to clear
-        if let Some(blocks) = self.lookup.get(&worker) {
-            let blocks_to_clear: Vec<_> = blocks.values().collect();
+    pub fn clear_all_blocks(&mut self, worker_id: WorkerId) {
+        // Collect all WorkerWithDpRank keys that match this worker_id
+        let workers_to_clear: Vec<WorkerWithDpRank> = self
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/indexer.rs]: Can add helper function that will collect dp workers under a given worker_id.

File: lib/llm/src/kv_router/indexer.rs [hunk 9]
```diff
@@ -671,10 +701,10 @@ impl OverlapScores {
     ///
     /// ### Arguments
     ///
-    /// * `workers` - An iterator over `WorkerId` references.
+    /// * `workers` - An iterator over `WorkerWithDpRank` references.
     pub fn update_scores<'a, I>(&mut self, workers: I)
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/indexer.rs:687]: Is this used? Can we delete if not.

File: lib/llm/src/kv_router/scheduler.rs [hunk 10]
```diff
@@ -117,14 +119,10 @@ impl KvScheduler {
...
 
-        let worker_ids: Vec<i64> = instances
-            .iter()
-            .map(|instance| instance.instance_id)
-            .collect();
         let slots = Arc::new(ActiveSequencesMultiWorker::new(
             component.clone(),
             block_size as usize,
-            worker_ids,
+            workers_with_configs.read().await.clone(),
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/scheduler.rs]: Add comment that this includes information about dp ranks, which is necessary now for the slot manager.

File: lib/llm/src/kv_router/scheduler.rs [hunk 11]
```diff
@@ -363,21 +360,21 @@ impl KvScheduler {
...
+        workers.extend(decode_blocks.keys().copied());
+        workers.extend(prefill_tokens.keys().copied());
 
         // Create PotentialLoad for each worker
         let mut loads = Vec::new();
-        for worker_id in worker_ids {
+        for worker in workers {
             loads.push(PotentialLoad {
-                worker_id,
+                worker_id: worker.worker_id,
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/scheduler.rs:374]: Why does this not include dp_rank in the key for PotentialLoad?

Comment (alec-flowers) [reply lib/llm/src/kv_router/scheduler.rs:374]: worker: WorkerwithDpRank

File: lib/llm/src/kv_router/scheduler.rs [hunk 12]
```diff
@@ -535,29 +546,30 @@ impl WorkerSelector for DefaultWorkerSelector {
...
-        let best_worker_id = softmax_sample(&worker_logits, temperature);
-        let best_logit = worker_logits[&best_worker_id];
+        let best_worker = softmax_sample(&worker_logits, temperature);
+        let best_logit = worker_logits[&best_worker];
 
-        let best_overlap = *overlaps.get(&best_worker_id).unwrap_or(&0);
+        let best_overlap = *overlaps.get(&best_worker).unwrap_or(&0);
         let total_blocks_info = workers
-            .get(&best_worker_id)
+            .get(&best_worker.worker_id)
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/scheduler.rs:559]: Write comment that total_block_info / total_kv_blocks is written to MDC as part of worker_id (so once for all dp_ranks)

File: lib/llm/src/kv_router/sequence.rs [hunk 13]
```diff
@@ -293,7 +293,7 @@ impl ActiveSequencesMultiWorker {
     pub fn new(
         component: Component,
         block_size: usize,
-        worker_ids: Vec<WorkerId>,
+        workers_with_configs: HashMap<i64, Option<ModelRuntimeConfig>>,
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/sequence.rs:296]: i64 -> WorkerId -> import from inside protocol.rs

File: lib/llm/src/kv_router/sequence.rs [hunk 14]
```diff
@@ -547,41 +555,53 @@ impl ActiveSequencesMultiWorker {
     }
 
     /// Update the set of workers, adding and removing as needed
-    pub fn update_workers(&self, new_worker_ids: Vec<WorkerId>) {
-        let current_workers: HashSet<WorkerId> =
+    pub fn update_workers(
+        &self,
+        new_workers_with_configs: HashMap<i64, Option<ModelRuntimeConfig>>,
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/sequence.rs:560]: workerID instead of i64

Comment (alec-flowers) [reply lib/llm/src/kv_router/sequence.rs:560]: Maybe create in protocols.rs an object for 
HashMap<WorkerId, Option<ModelRuntimeConfig>>

File: lib/llm/src/kv_router/sequence.rs [hunk 15]
```diff
@@ -1087,7 +1130,7 @@ mod tests {
                 None, // No token sequence
                 12,   // ISL (12 tokens)
                 0,    // no overlap
-                0,    // worker_id
+                WorkerWithDpRank::from_worker_id(0),
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/sequence.rs:1151]: This just seems like we are testing only different worker_ids.

Can we have a test or two where we actually add different dp ranks?

File: lib/llm/src/local_model/runtime_config.rs [hunk 16]
```diff
@@ -34,6 +38,25 @@ pub struct ModelRuntimeConfig {
     pub tensor_model_config: Option<tensor::TensorModelConfig>,
 }
 
+const fn default_data_parallel_size() -> u32 {
+    1
+}
+
+impl Default for ModelRuntimeConfig {
```
Comment (alec-flowers) [inline lib/llm/src/local_model/runtime_config.rs:45]: Why do we need to explicitly write out this default? Shouldn't this be handled by serdes default and the notes that you put on serde?

Comment (alec-flowers) [reply lib/llm/src/local_model/runtime_config.rs:45]: Serde only handles deserialization, not the Default trait itself.

File: lib/runtime/src/utils/worker_monitor.rs [hunk 17]
```diff
@@ -26,34 +26,59 @@ struct LoadEvent {
...
 }
 
-/// Worker load monitoring state
-#[derive(Clone, Debug)]
+/// Worker load monitoring state per dp_rank
+#[derive(Clone, Debug, Default)]
 pub struct WorkerLoadState {
-    pub kv_active_blocks: Option<u64>,
-    pub kv_total_blocks: Option<u64>,
+    pub kv_active_blocks: HashMap<u32, u64>,
```
Comment (alec-flowers) [inline lib/runtime/src/utils/worker_monitor.rs:52]: Can we make a dp_rank = u32 in the protocol.rs ?

File: lib/bindings/c/src/lib.rs [hunk 18]
```diff
@@ -207,6 +207,7 @@ fn kv_event_create_stored_from_parts(
             parent_hash: kv_params.parent_hash.map(ExternalSequenceBlockHash),
         }),
         event_id: kv_params.event_id,
+        dp_rank: 0,
```
Comment (alec-flowers) [inline lib/bindings/c/src/lib.rs:210]: I think we can nuke the c/src file. We don't use the c bindings anymore.

File: lib/bindings/python/rust/llm/kv.rs [hunk 19]
```diff
@@ -314,8 +324,13 @@ pub(crate) struct OverlapScores {
 #[pymethods]
 impl OverlapScores {
     #[getter]
-    fn scores(&self) -> HashMap<llm_rs::kv_router::indexer::WorkerId, u32> {
-        self.inner.scores.clone()
+    fn scores(&self) -> HashMap<(i64, u32), u32> {
```
Comment (alec-flowers) [inline lib/bindings/python/rust/llm/kv.rs:327]: Can we name these instead of - HashMap<(i64, u32), u32>

File: lib/bindings/python/rust/llm/kv.rs
Comment (alec-flowers) [inline lib/bindings/python/rust/llm/kv.rs:1114]: Create another function called
best_worker

Add deprecation notice to this function that it is being deprecated in favor of the best_worker() function (which will return dp_rank)

File: lib/bindings/python/rust/llm/kv.rs [hunk 20]
```diff
@@ -1130,9 +1152,24 @@ impl KvPushRouter {
                 .await
                 .map_err(to_pyerr)?;
 
+            // Aggregate loads over dp_ranks - sum up loads for the same worker_id
+            let mut aggregated: HashMap<i64, llm_rs::kv_router::scheduler::PotentialLoad> =
```
Comment (alec-flowers) [inline lib/bindings/python/rust/llm/kv.rs]: Can we label i64

File: lib/bindings/python/rust/llm/kv.rs
Comment (alec-flowers) [inline lib/bindings/python/rust/llm/kv.rs:1187]: This also needs a deprecation notice and a new function.

File: lib/llm/src/kv_router/approx.rs [hunk 21]
```diff
@@ -327,7 +329,7 @@ impl ApproxKvIndexer {
     pub async fn process_routing_decision_for_request(
```
Comment (alec-flowers) [inline lib/llm/src/kv_router/approx.rs:329]: Potential to add e2e test with mocker for the approximate router

**Review by alec-flowers — APPROVED** [review]
Left a few comments, but nothing blocking. Great work!

**Comment by alec-flowers** [comment]
We want to update the trtllm and sglang python to make use of this.

That requires figuring out the event publishing for multiple dp worker for trtllm and sglang. We also need to pass the dp_rank  to the API so that the frameworks handle the routing. 

Future:
- Router that will balance the dp_workers to avoid bubbles. Special router cost function for dp routing.

**Review by nv-tusharma — APPROVED** [review]

File: lib/bindings/c/src/lib.rs (hunk 18 above)
Comment (PeaBrane) [reply lib/bindings/c/src/lib.rs:210]: leaving this for future

**Comment by michaelfeil** [comment]
@PeaBrane Concern: Actually in trtllm you will have 1 publisher class, but it receives and processes events for all mpi ranks.
